File: frontend/src/services/stateHubService.js
import api from './api';
import { API_ROUTES } from '../constants/apiRoutes';
import { normalizeResponse } from '../utils/normalizeResponse';

export const stateHubService = {
  getAll: async () => {
    const response = await api.get(`${API_ROUTES.STATE_HUBS}/`);
    return normalizeResponse(response.data);
  },

  create: async (payload) => {
    const response = await api.post(`${API_ROUTES.STATE_HUBS}/`, payload);
    return normalizeResponse(response.data);
  },

  update: async (id, payload) => {
    const response = await api.put(`${API_ROUTES.STATE_HUBS}/${id}`, payload);
    return normalizeResponse(response.data);
  },

  delete: async (id) => {
    const response = await api.delete(`${API_ROUTES.STATE_HUBS}/${id}`);
    return normalizeResponse(response.data);
  },
};

File: backend/app/services/state_hub_service.py
from sqlalchemy.orm import Session
from app.models.schema import StateHub

class StateHubService:
    @staticmethod
    def get_all(db: Session, company_id: int):
        return db.query(StateHub).filter(StateHub.company_id == company_id).all()

    @staticmethod
    def get_all_across_companies(db: Session):
        return db.query(StateHub).all()

    @staticmethod
    def get_by_id(db: Session, hub_id: int, company_id: int):
        return db.query(StateHub).filter(StateHub.id == hub_id, StateHub.company_id == company_id).first()

    @staticmethod
    def create(db: Session, company_id: int, data: dict):
        hub = StateHub(company_id=company_id, **data)
        db.add(hub)
        db.flush()
        db.refresh(hub)
        return hub

    @staticmethod
    def update(db: Session, hub_id: int, company_id: int, data: dict):
        hub = StateHubService.get_by_id(db, hub_id, company_id)
        if not hub:
            raise ValueError("State Hub not found")
        for key, value in data.items():
            setattr(hub, key, value)
        db.flush()
        db.refresh(hub)
        return hub

    @staticmethod
    def delete(db: Session, hub_id: int, company_id: int):
        hub = StateHubService.get_by_id(db, hub_id, company_id)
        if not hub:
            raise ValueError("State Hub not found")
        if hub.warehouses:
            raise ValueError("Cannot delete State Hub with assigned warehouses")
        db.delete(hub)
        db.flush()
        return True
